File: solutions/Leetcode_890/leetcode_890.cc
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;


class Solution1
{
public:
    vector<string> findAndReplacePattern(vector<string>& words, string pattern)
    {
        string p = normalize(pattern);
        vector<string> ret;
        for (string word : words)
        {
            string normalized_word = normalize(word);
            if (normalized_word == p)
                ret.push_back(word);
        }
        return ret;
    }

private:
    string normalize(string word)
    {
        unordered_map<char, int> mapping;
        for (int i = 0; i < word.size(); i++)
        {
            if (mapping.find(word[i]) == mapping.end())
                mapping[word[i]] = mapping.size();
            word[i] = 'a' + mapping[word[i]];
        }
        return word;
    }
};

File: solutions/Leetcode_890/leetcode_890.py
from typing import List

class Solution1:
    def find_and_replace_pattern(self, words, pattern):
        def normalize(word):
            mapping = {}
            for x in word:
                yield mapping.setdefault(x, len(mapping))
        p = list(normalize(pattern))
        return [word for word in words if all(x == y for x, y in zip(normalize(word), p))]

class SolutionMINE:
    def find_and_replace_pattern(self, words: List[str], pattern: str) -> List[str]:
        mapping = {}
        appeared = []
        ret = []
        for word in words:
            mapping.clear()
            appeared.clear()
            for i in range(len(word)):
                if pattern[i] not in mapping:
                    if word[i] not in appeared:
                        mapping[pattern[i]] = word[i]
                        appeared.append(word[i])
                    else:
                        break
                if mapping[pattern[i]] != word[i]:
                    break
                if i == len(word) -1:
                    ret.append(word)
        return ret

if __name__ == '__main__':
    words = ["a","b","c"]
    pattern = 'a'
    ans = Solution1().find_and_replace_pattern(words, pattern)
    print(ans)